src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
 }
 
-struct TxValues {
+class TxValues {
+public:
     CAmount t_inputs_total{0};
     CAmount z_inputs_total{0};
     CAmount t_outputs_total{0};
     CAmount z_outputs_total{0};
     CAmount targetAmount{0};
-    bool selectedUTXOCoinbase{false};
+
+    CAmount GetChangeAmount() {
+        return t_inputs_total + z_inputs_total - targetAmount;
+    }
 };
 
+bool IsFromAnyTaddr(const PaymentSource& paymentSource) {
+    return std::visit(match {
+        [&](const FromAnyTaddr& fromAny) {
+            return true;
+        },
+        [&](const PaymentAddress& addr) {
+            return false;
+        }
+    }, paymentSource);
+}
+
+// Construct and send the transaction, returning the resulting txid.
+// Errors in transaction construction will throw.
+//
 // Notes:
 // 1. #1159 Currently there is no limit set on the number of joinsplits, so size of tx could be invalid.
 // 2. #1360 Note selection is not optimal
 // 3. #1277 Spendable notes are not locked, so an operation running in parallel could also try to use them
-bool AsyncRPCOperation_sendmany::main_impl() {
-
+uint256 AsyncRPCOperation_sendmany::main_impl() {
+    // TODO UA: this check will become meaningless.
+    bool isfromzaddr_ = isfromsprout_ || isfromsapling_;
     assert(isfromtaddr_ != isfromzaddr_);
 
-    bool isSingleZaddrOutput = (t_outputs_.size()==0 && z_outputs_.size()==1);
-    bool isMultipleZaddrOutput = (t_outputs_.size()==0 && z_outputs_.size()>=1);
-    bool isPureTaddrOnlyTx = (isfromtaddr_ && z_outputs_.size() == 0);
-    CAmount minersFee = fee_;
     TxValues txValues;
 
     // First calculate the target
-    for (SendManyRecipient & t : t_outputs_) {
-        txValues.t_outputs_total += t.amount;
-    }
-
-    for (SendManyRecipient & t : z_outputs_) {
-        txValues.z_outputs_total += t.amount;
+    int shieldedRecipients = 0;
+    int transparentRecipients = 0;
+    for (const SendManyRecipient& recipient : recipients_) {
+        std::visit(match {
+            [&](const CKeyID& addr) {
+                transparentRecipients += 1;
+                txValues.t_outputs_total += recipient.amount;
+            },
+            [&](const CScriptID& addr) {
+                transparentRecipients += 1;
+                txValues.t_outputs_total += recipient.amount;
+            },
+            [&](const libzcash::SproutPaymentAddress& addr) {
+                // unreachable; currently disallowed by checks at construction
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Sending to Sprout is disabled.");
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) {
+                txValues.z_outputs_total += recipient.amount;
+                shieldedRecipients += 1;
+                if (isfromsprout_ && !allowRevealedAmounts_) {
+                    throw JSONRPCError(
+                            RPC_INVALID_PARAMETER,
+                            "Sending from Sprout to Sapling is not enabled by default because it will "
```

Comment:
```suggestion
                            "Sending between shielded pools is not enabled by default because it will "
```
We should make the error message reflect what the generalized policy will be, since we already know it.